### ISSUE TITLE
Undeletable Old User Data in anope_db_ChanAccess Despite Account Removal

### ISSUE THREAD

**Issue by Madriix** [anope/anope#451]
### Description

Old user account data remains in the anope_db_ChanAccess table and cannot be deleted, even though the accounts have been removed or have expired.

### Steps to reproduce the issue

1. The user creates an Anope account (in my case, users are created using a table from the sql_authentication module).
2. Sometimes, for example, if we remove a user, they still remain in the anope_db_ChanAccess table. It doesn’t always delete the data.
No idea if the user '_Sissi' was removed from Anope due to expiration or a drop (I think it was expiration).

### Describe the results you received

For example, there’s currently an account named '_Sissi' that no longer exists (it did exist previously), but it’s still present in the anope_db_ChanAccess.mask table (it shouldn’t be there).

### Describe the results you expected

The SQL table(s) should not retain any old data if an account is deleted or has expired. However, it’s important to keep the 'setters' but not the 'access' entries where they appear, as these accounts no longer exist.

### Anope version

2.1.10

**Comment by SadieCat** [comment]
I can't reproduce this locally which is odd.

**Comment by SadieCat** [comment]
How have you restarted your services recently? I'm wondering if there was a delete queued and it was unable to finish flushing it before the process was terminated.

**Comment by Madriix** [comment]
@SadieCat 
I just reviewed the Anope tables' engines and noticed that one table, named `anope_chan`, is set to InnoDB. All other Anope tables are using MyISAM. I'm not sure what the Anope standard is for table engines, and I can't delete and recreate them all since there are many accounts (around 6,000 in `anope_db_NickCore`).

Each time I start Anope, I do so via SSH with the command: `cd /home/irc/anope2/services/bin` and then `./anope`. To stop Anope, I always use the command: `/msg OperServ SHUTDOWN`.

I'm using the `sql_authentication` module to allow users to authenticate with a table named `compte`. Anope is installed on the same dedicated Debian 12 server as UnrealIRCd, and everything is set up via localhost/127.0.0.1.

It’s worth noting that these SQL tables have been in place since at least 2019 and have never been deleted. The table collations vary, but I'm not sure which one to standardize on. Currently, there are some using `latin1_swedish_ci` (which was the default in 2019), and a few others using `utf8mb4_general_ci` and `utf8mb3_general_ci`.

In the Anope configuration, there's a module named `db_sql and db_sql_live`. It’s enabled on `db_sql`, with the prefix `anope_db_` also enabled, and the import setting is set to `true` (always on). The `db_flatfile` option is also enabled, with all default settings.

**Comment by SadieCat** [comment]
Any tables using MyISAM or a character set other than utf8mb4are outdated but unfortunately there's no way to migrate them whilst online. Its possible that using MyISAM is causing performance issues here as it is an obsolete storage engine which is known to have worse performance. I'd recommend trying the following steps:

0. Run `/msg OperServ SHUTDOWN`
1. Back up your SQL anope_db_* SQL tables
2. Update your config to load db_flatfile after db_sql(_live)
3. Start Anope
4. Run `/msg OperServ SHUTDOWN` (this should write the flatfile database)
5. Drop all of your anope_db_* SQL tables (make sure you backed them all up in step 1)
6. Load db_sql with the "import" option enabled (note: db_sql_live can't handle imports).
7. Start Anope and wait while it imports your database with the latest schema
8. Run `/msg OperServ SHUTDOWN`
9. Restore your config to how it was to start with

**Comment by SadieCat** [comment]
>In the Anope configuration, there's a module named db_sql and db_sql_live. It’s enabled on db_sql, with the prefix anope_db_ also enabled, and the import setting is set to true (always on). The db_flatfile option is also enabled, with all default settings.

I just noticed this. You shouldn't have the import option enabled after you have imported your database as that just requeues a bunch of INSERTs that don't need to be executed. Once you follow the steps I mentioned in my previous post you should set this to false.

**Comment by Madriix** [comment]
@SadieCat on step 1 (Back up your anope_db_* SQL tables) should I also delete the 'anope_' tables without the 'db' prefix, like anope_user and anope_chan? Or should I not?

**Comment by SadieCat** [comment]
Those tables are from irc2sql or chanstats and aren't related to your services database so you should leave them alone.

**Comment by Madriix** [comment]
@SadieCat He did a ping timeout:

> ERROR: Closing Link: [127.0.0.1] (Registration Timeout)

I restarted `./anope`, is that correct too?
It recreated the tables before the ping timeout

> Update your config to load db_flatfile after db_sql(_live)

On step 2, I have to deactivate the db_live(live) module by adding # in front of module?

**Comment by SadieCat** [comment]
>ERROR: Closing Link: [127.0.0.1] (Registration Timeout)

You might need to raise your ping timeout for the server for the initial import (this is a known bug) if it doesn't import the data before this happens.

> On step 2, I have to deactivate the db_live(live) module by adding # in front of module?

You need to have db_sql (or db_sql_live) loaded at the same time as db_flatfile so it can export to flatfile.

**Comment by Madriix** [comment]
I have to change the direction of the modules, put the db_flatfile module below sql dq_sql?
On step 2: Update your config to load db_flatfile after db_sql(_live)

**Comment by SadieCat** [comment]
I'm not sure why you have both loaded currently but that will probably also work. As long as you have a full flatfile database ready before the import.

**Comment by Madriix** [comment]
@SadieCat Yes, I've always had both `db_flatfile` and `db_sql` at the same time. I followed all the steps, but at step 7, there’s always a ping timeout, so I restart `./anope`, and then it's fine. In the UnrealIRCd configuration, I increased the `pingfreq` for the Anope server block, but it didn't resolve the ping timeout issue; it still occurs. I noticed that only the `anope_db_NickAlias` table loads 4 or 5 aliases of connected users. Why not all of them? I’ve often seen this issue before, but I still can't understand it.

For the `ChanAccess` table, I resolved the issue by using the `anope.db` from the backup folder dated from last night, and that’s how the problem was resolved.

**Comment by SadieCat** [comment]
You may also need to temporarily unload chanstats and/or irc2sql as they may be causing an excessive amount of inserts on link.

**Comment by Madriix** [comment]
@SadieCat This didn't change anything

**Comment by H7-25** [comment]
This seems an casemapping issue, are you sure acccount you can't delete were not created with a different casemapping in the config ?

	/*
	 * The case mapping used by services. This must be set to a valid locale name
	 * installed on your machine. Services use this case map to compare, with
	 * case insensitivity, things such as nick names, channel names, etc.
	 *
	 * We provide two special casemaps shipped with Anope, ascii and rfc1459.
	 *
	 * This value should be set to what your IRCd uses, which is probably rfc1459,
	 * however Anope has always used ascii for comparison, so the default is ascii.
	 *
	 * Changing this value once set is not recommended.
	 */
	casemap = "rfc1459"

**Comment by SadieCat** [comment]
@H7-25 I don't think this is a casemap issue as that nickname shouldn't be casemapped.

@Madriix Can you try again using the latest 2.1 Git? I made an improvement that I think should fix the SQL import.

**Comment by Madriix** [comment]
@SadieCat I can modify the init.cpp file and just put this "SocketEngine::Process();" to save time? or do I have to update completely with git? this last one will be super long

**Comment by SadieCat** [comment]
You can try that but there's been some other database module changes since the last 2.1 release so I can't be sure if it will work the same as it did for me.

**Comment by Madriix** [comment]
@SadieCat I recompiled, there was no ping timeout this time. For the anope_db_NickAlias ​​table I don't understand, it doesn't load all the aliases Where could the problem come from?

Yet in anope.db they are present, see here:


> OBJECT NickAlias
> DATA nick Glow3
> DATA last_quit 
> DATA last_realname unknown
> DATA last_usermask *@*
> DATA last_realhost 
> DATA time_registered 1731437051
> DATA last_seen 1731437051
> DATA nc Glow

I created the alias Glow2, it does not appear in the tables, there are more than 2000 aliases, but there are only 6 (accounts) that appear in the anope_db_NickAlias ​​table. It seems that only this table is causing problems

**Comment by Madriix** [comment]
I installed the Anope Git this morning and I'm still at the same point as my previous message: no ping timeout, but there is an issue with the anope_db_NickAlias table. It was created, but wasn't filled correctly; around 2,000 to 5,000 aliases are missing. Only 6 were added this morning. I've been having problems with this table for a long time, even with Anope 2.0

**Comment by Madriix** [comment]
@SadieCat I ran `./anope -support` and got this error:

```
[Nov 13 17:36:22.200798 2024] Debug: Error executing query INSERT INTO `anope_db_ChannelInfo` (`id`,`BS_FANTASY`,`BS_GREET`,`CS_NO_EXPIRE`,`CS_STATS` ...) Column 'BS_GREET' cannot be null
```

In my SQL table, I see the columns in this order: `id | timestamp | BS_FANTASY | CS_NO_EXPIRE | CS_STATS`.
BS_GREET column is missing

I will create it to see

**Comment by Madriix** [comment]
@SadieCat 
There is also this error when creating an alias: 

> [Nov 13 17:59:45.144342 2024] Debug: Error executing query INSERT INTO `anope_db_NickAlias` (`id`,`NS_NO_EXPIRE`,`last_quit`,`last_realhost`,`last_realname`,`last_seen`,`last_usermask`,`nc`,`nick`,`time_registered`,`vhost_creator`,`vhost_host`,`vhost_ident`,`vhost_time`) VALUES (0,NULL,NULL,NULL,'unknown',1731517183,'*@*','Glow','testy',1731517183,NULL,NULL,NULL,NULL) ON DUPLICATE KEY UPDATE `NS_NO_EXPIRE`=........name`=VALUES(`name`),`repeattimes`=VALUES(`repeattimes`),`time_registered`=VALUES(`time_registered`),`ttb`=VALUES(`ttb`): Column 'CS_STATS' cannot be null

**Comment by Madriix** [comment]
Ah oups, j'me suis trompé, j'ai coupé une ligne qu'était en rapport avec ChannelInfo, voici la vraie erreur pour les aliases: 

```[Nov 13 17:59:45.144342 2024] Debug: Error executing query INSERT INTO `anope_db_NickAlias` (`id`,`NS_NO_EXPIRE`,`last_quit`,`last_realhost`,`last_realname`,`last_seen`,`last_usermask`,`nc`,`nick`,`time_registered`,`vhost_creator`,`vhost_host`,`vhost_ident`,`vhost_time`) VALUES (0,NULL,NULL,NULL,'unknown',1731517183,'*@*','Glow','testy',1731517183,NULL,NULL,NULL,NULL) ON DUPLICATE KEY UPDATE `NS_NO_EXPIRE`=VALUES(`NS_NO_EXPIRE`),`last_quit`=VALUES(`last_quit`),`last_realhost`=VALUES(`last_realhost`),`last_realname`=VALUES(`last_realname`),`last_seen`=VALUES(`last_seen`),`last_usermask`=VALUES(`last_usermask`),`nc`=VALUES(`nc`),`nick`=VALUES(`nick`),`time_registered`=VALUES(`time_registered`),`vhost_creator`=VALUES(`vhost_creator`),`vhost_host`=VALUES(`vhost_host`),`vhost_ident`=VALUES(`vhost_ident`),`vhost_time`=VALUES(`vhost_time`): Column 'NS_NO_EXPIRE' cannot be null```

**Comment by SadieCat** [comment]
Please let me know if you encounter any data hanging about in the database with the recent changes.

**Comment by Madriix** [comment]
@SadieCat The tables are recreated at the first start of anope just after deleting the tables, and all this with import = true. I don't see any problem with recording, everything seems complete when creating the tables

I have to create a new topic to report a problem with import = true

**Comment by SadieCat** [comment]
Have you had any more problems with this?

**Comment by Madriix** [comment]
hi @SadieCat no, I don’t think so. However, I’m going off-topic here, but with the latest update 2.1.11, you sometimes need to use /OS STATS reset 2 or 3 times because the statistics won’t reset to zero. Sometimes they do reset, but if I use /OS STATS again, it just restores the previous stats from before the reset. It seems to have been happening since the beginning of December
